### PR TITLE
Add HTTPS support to Open_uri

### PR DESCRIPTION
**Pull request by robhoes**
This required a bunch of libraries to be split off and moved around (hopefully for the better).

### CODE REVIEW

File: ocaml/xapi-idl/lib/dune [hunk 1]
```diff
@@ -12,6 +12,7 @@
    message-switch-unix 
    mtime
    mtime.clock.os
+   open-uri
```
Comment (edwintorok) [inline ocaml/xapi-idl/lib/dune:15]: for consistency should it be `xapi-open-uri`?

File: ocaml/libs/open-uri/open_uri.ml [hunk 2]
```diff
@@ -53,6 +53,20 @@ let with_open_uri uri f =
              (Uri.to_string uri)
           )
   )
+  | Some "https" -> (
+    match (Uri.host uri, Uri.port uri) with
+    | Some host, Some port ->
+        Stunnel.with_connect ~verify_cert:(Stunnel_client.pool ()) host port (fun s ->
+          f Safe_resources.Unixfd.(!(s.Stunnel.fd)))
```
Comment (edwintorok) [inline ocaml/libs/open-uri/open_uri.ml:60]: that is not really a safe way to pass that fd, the wrapper might decide to close it at some point, can we instead pass it the `s.Stunnel.fd` directly and let it do the unwrapping internally?

**Comment by edwintorok** [comment]
Looks like a `make format` is needed after the `Uuidx` changes, it probably changed the length of some lines just enough to trigger splitting.

File: ocaml/xapi-idl/lib/dune (hunk 1 above)
Comment (robhoes) [reply ocaml/xapi-idl/lib/dune:15]: Hmm yes, that's probably better.

File: ocaml/libs/open-uri/open_uri.ml (hunk 2 above)
Comment (robhoes) [reply ocaml/libs/open-uri/open_uri.ml:60]: Do you mean that `Safe_resources.Unixfd.(!)` will close it?

I'll take a look, but it comes down to the semantics of the existing function, and the way `f` is used by callers now. If we'd push down the Safe_resources thing, then we'll also need to do that for the other cases.

File: ocaml/libs/open-uri/open_uri.ml [hunk 3]
```diff
@@ -53,6 +53,20 @@ let with_open_uri uri f =
              (Uri.to_string uri)
           )
   )
+  | Some "https" -> (
+    match (Uri.host uri, Uri.port uri) with
+    | Some host, Some port ->
+        Stunnel.with_connect ~verify_cert:(Stunnel_client.pool ()) host port (fun s ->
```
[the remark this answers is not in the document]
Comment (robhoes) [reply ocaml/libs/open-uri/open_uri.ml]: I've now added `?verify_cert`. It's a bit awkward, as the intention of this function is to be guided by the URL alone, and now takes an argument that is only relevant for one of the cases after parsing the URL. But is seems the best we can do now.

File: ocaml/libs/open-uri/open_uri.ml (hunk 2 above)
Comment (robhoes) [reply ocaml/libs/open-uri/open_uri.ml]: The best way forward seems to be to adopt Safe_resources for all fds in this function, but that would be a much bigger change. At least `Stunnel.with_connect` does not close the fd as far as I can see.

File: ocaml/xapi-idl/lib/dune (hunk 1 above)
Comment (robhoes) [reply ocaml/xapi-idl/lib/dune]: Done.

**Comment by psafont** [comment]
This is a reminder for myself to push the new .opam files to xs-opam when this is merged

File: ocaml/libs/open-uri/open_uri.ml [hunk 4]
```diff
@@ -53,6 +53,25 @@ let with_open_uri uri f =
              (Uri.to_string uri)
           )
   )
+  | Some "https" -> (
+      let verify_cert =
+        Option.value ~default:(Stunnel_client.pool ()) verify_cert
```
Comment (psafont) [inline ocaml/libs/open-uri/open_uri.ml:58]: I like that users are forced to be explicit about the verification they need if they require something else other than pool verification

**Review by psafont — APPROVED** [review]
Changes look good. I think this change needs a build to test and flush out potential issues with open_uris using non-pool hosts

**Comment by robhoes** [comment]
@psafont Thanks. Note that the changes to Open_uri here actually don't yet take effect anywhere, as it is not yet called with `https` URLs. That will be next :)

**Review by lindig — APPROVED** [review]
99% of all changes are Uuid/Uuidx. This is best reviewed by commits.

**Pull request merged**